crates/biome_graphql_parser/src/lexer/mod.rs

```diff
@@ -0,0 +1,720 @@
+            b"QUERY" => UPPER_QUERY_KW,
+            b"MUTATION" => UPPER_MUTATION_KW,
+            b"SUBSCRIPTION" => UPPER_SUBSCRIPTION_KW,
+            b"FIELD" => UPPER_FIELD_KW,
+            b"FRAGMENT_DEFINITION" => FRAGMENT_DEFINITION_KW,
+            b"FRAGMENT_SPREAD" => FRAGMENT_SPREAD_KW,
+            b"INLINE_FRAGMENT" => INLINE_FRAGMENT_KW,
+            b"VARIABLE_DEFINITION" => VARIABLE_DEFINITION_KW,
+            b"SCHEMA" => UPPER_SCHEMA_KW,
+            b"SCALAR" => UPPER_SCALAR_KW,
+            b"OBJECT" => UPPER_OBJECT_KW,
+            b"FIELD_DEFINITION" => FIELD_DEFINITION_KW,
+            b"ARGUMENT_DEFINITION" => ARGUMENT_DEFINITION_KW,
+            b"INTERFACE" => UPPER_INTERFACE_KW,
+            b"UNION" => UPPER_UNION_KW,
+            b"ENUM" => UPPER_ENUM_KW,
+            b"ENUM_VALUE" => ENUM_VALUE_KW,
+            b"INPUT_OBJECT" => INPUT_OBJECT_KW,
+            b"INPUT_FIELD_DEFINITION" => INPUT_FIELD_DEFINITION_KW,
+            _ => GRAPHQL_NAME,
+        }
+    }
+
+    #[inline]
+    fn consume_unexpected_character(&mut self) -> GraphqlSyntaxKind {
+        self.assert_current_char_boundary();
+
+        let char = self.current_char_unchecked();
+        let err = ParseDiagnostic::new(
+            format!("unexpected character `{}`", char),
+            self.text_position()..self.text_position() + char.text_len(),
+        );
+        self.diagnostics.push(err);
+        self.advance(char.len_utf8());
+
+        ERROR_TOKEN
+    }
+
+    fn consume_number(&mut self, first: u8) -> GraphqlSyntaxKind {
+        self.assert_current_char_boundary();
+
+        let start = self.text_position();
+
+        let mut state = match first {
+            b'-' => LexNumberState::Minus,
+            b'0' => LexNumberState::LeadingZero,
+            _ => LexNumberState::IntegerPart,
+        };
+        self.advance(1);
+
+        while let Some(chr) = self.current_byte() {
+            let new_state = match chr {
+                b'0'..=b'9' => self.consume_digit(chr, state),
+                b'.' => self.consume_fraction(chr, state),
+                b'e' | b'E' => self.consume_exponent(chr, start, state),
+                _ => break,
+            };
+            state = new_state;
+        }
+
+        match state {
+            LexNumberState::LeadingZero | LexNumberState::IntegerPart => GRAPHQL_INT_LITERAL,
+            LexNumberState::FractionalPart | LexNumberState::Exponent => GRAPHQL_FLOAT_LITERAL,
+            LexNumberState::Minus => {
+                self.diagnostics.push(ParseDiagnostic::new(
+                    "Unexpected token `-`",
+                    start..self.text_position(),
+                ));
+
+                ERROR_TOKEN
+            }
+            LexNumberState::Invalid(diagnostic) => {
+                self.diagnostics.push(diagnostic);
+                ERROR_TOKEN
+            }
+        }
+    }
+
+    fn consume_digit(&mut self, chr: u8, state: LexNumberState) -> LexNumberState {
```

Comment:
What's the difference between a digit and a number? It's worth adding a comment that briefly explains it. Also, `LexNumberState` should be called `LexDigitState`, to avoid confusion.